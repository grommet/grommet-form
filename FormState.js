'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _utils = require('./utils');

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var FormState = function FormState(rules) {
  var obj = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var onChange = arguments[2];

  _classCallCheck(this, FormState);

  _initialiseProps.call(this);

  if (!rules || typeof rules !== 'function' && !Object.keys(rules).length) {
    throw new Error('Rules is a required argument');
  }
  this.rules = rules;
  this.obj = JSON.parse(JSON.stringify(obj));
  this.onChange = onChange;
};

var _initialiseProps = function _initialiseProps() {
  var _this = this;

  this.set = function (key, value) {
    if (typeof key === 'string') {
      (0, _utils.setValueByKey)(_this.obj, key, value);
    } else if ((typeof key === 'undefined' ? 'undefined' : _typeof(key)) === 'object') {
      _this.obj = (0, _utils.deepMerge)(_this.obj, key);
    }

    if (_this.onChange) {
      _this.onChange(_defineProperty({}, key, value));
    }
  };

  this.get = function () {
    return _extends({}, _this.obj);
  };

  this.getResource = function () {
    return JSON.parse(JSON.stringify(_this.obj).replace(/"\s+|\s+"/g, '"'));
  };

  this.getErrors = function (rules, target) {
    var errors = {};
    var targetRules = rules || (typeof _this.rules === 'function' ? _this.rules(_this.obj) : _this.rules);
    var targetObject = target || _this.obj;
    Object.keys(targetRules).forEach(function (key) {
      var rule = targetRules[key];
      var processedValue = typeof targetObject[key] === 'string' ? targetObject[key].trim() : targetObject[key];
      if (typeof rule === 'function') {
        var value = (0, _utils.getValueByKey)(targetObject, key);
        if (Array.isArray(value)) {
          var errorsArray = [];
          value.forEach(function (v, index) {
            var message = rule(value, index, targetObject);
            if (message) {
              errorsArray[index] = message;
            }
          });
          if (errorsArray.length) {
            (0, _utils.setValueByKey)(errors, key, errorsArray);
          }
        } else {
          var message = rule(value, undefined, targetObject);
          if (message) {
            errors[key] = message;
          }
        }
      } else if ((0, _utils.isObject)(rule)) {
        var ruleErrors = _this.getErrors(rule, targetObject[key]);
        if (Object.keys(ruleErrors).length) {
          errors[key] = ruleErrors;
        }
      } else if (typeof rule === 'string' && (!processedValue || processedValue.length <= 0)) {
        (0, _utils.setValueByKey)(errors, key, rule);
      }
    });
    return errors;
  };

  this.isValid = function () {
    return !Object.keys(_this.getErrors()).length;
  };
};

exports.default = FormState;