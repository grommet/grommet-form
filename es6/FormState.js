'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _utils = require('./utils');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var FormState = function FormState(rules) {
  var obj = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var onChange = arguments[2];

  _classCallCheck(this, FormState);

  _initialiseProps.call(this);

  if (!rules || typeof rules !== 'function' && !Object.keys(rules).length) {
    throw new Error('Rules is a required argument');
  }
  this.rules = rules;
  this.obj = obj;
  this.onChange = onChange;
};

var _initialiseProps = function _initialiseProps() {
  var _this = this;

  this.set = function (key, value) {
    if (typeof key === 'string') {
      (0, _utils.setValueByKey)(_this.obj, key, value);
    } else if ((typeof key === 'undefined' ? 'undefined' : _typeof(key)) === 'object') {
      _this.obj = (0, _utils.deepMerge)(_this.obj, key);
    }

    if (_this.onChange) {
      _this.onChange();
    }
  };

  this.get = function () {
    return _this.obj;
  };

  this.getErrors = function (rules, target) {
    var errors = {};
    var targetRules = rules || (typeof _this.rules === 'function' ? _this.rules(_this.obj) : _this.rules);
    var targetObject = target || _this.obj;
    Object.keys(targetRules).forEach(function (key) {
      var rule = targetRules[key];
      if (typeof rule === 'function') {
        var value = targetObject[key];
        if (Array.isArray(value)) {
          var errorsArray = [];
          value.forEach(function (v, index) {
            var message = rule(targetObject[key], index, targetObject);
            if (message) {
              errorsArray[index] = message;
            }
          });
          if (errorsArray.length) {
            errors[key] = errorsArray;
          }
        } else {
          var message = rule(targetObject[key], undefined, targetObject);
          if (message) {
            errors[key] = message;
          }
        }
      } else if ((0, _utils.isObject)(rule)) {
        var ruleErrors = _this.getErrors(rule, targetObject[key]);
        if (Object.keys(ruleErrors).length) {
          errors[key] = ruleErrors;
        }
      } else if (typeof rule === 'string' && !targetObject[key]) {
        errors[key] = rule;
      }
    });
    return errors;
  };

  this.isValid = function () {
    return !Object.keys(_this.getErrors()).length;
  };
};

exports.default = FormState;