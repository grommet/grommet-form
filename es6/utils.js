'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.isObject = isObject;
exports.deepMerge = deepMerge;
exports.deepCompare = deepCompare;
exports.getValueByKey = getValueByKey;
exports.setValueByKey = setValueByKey;
function isObject(item) {
  return item && (typeof item === 'undefined' ? 'undefined' : _typeof(item)) === 'object' && !Array.isArray(item);
}

function deepMerge(target) {
  for (var _len = arguments.length, sources = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    sources[_key - 1] = arguments[_key];
  }

  if (!sources.length) {
    return target;
  }
  // making sure to not change target (immutable)
  var output = _extends({}, target);
  var source = sources.shift();
  if (isObject(output) && isObject(source)) {
    Object.keys(source).forEach(function (key) {
      if (isObject(source[key])) {
        if (!output[key]) {
          output[key] = _extends({}, source[key]);
        } else {
          output[key] = deepMerge({}, output[key], source[key]);
        }
      } else {
        output[key] = source[key];
      }
    });
  }
  return deepMerge.apply(undefined, [output].concat(sources));
}

function deepCompare(obj1, obj2) {
  if (obj1 && obj2) {
    var diff = Object.keys(obj1).find(function (key) {
      if (obj1.hasOwnProperty(key) !== obj2.hasOwnProperty(key)) {
        return true;
      }
      switch (_typeof(obj1[key])) {
        case 'object':
          if (!deepCompare(obj1[key], obj2[key])) {
            return true;
          }
          break;
        case 'function':
          if (typeof obj2[key] === 'undefined' || obj1[key].toString() !== obj2[key].toString()) {
            return true;
          }
          break;
        default:
          if (obj1[key] !== obj2[key]) {
            return true;
          }
      }
      return false;
    });
    if (diff) {
      return false;
    }
    // check if obj2 has any props that do not exist in obj1
    diff = Object.keys(obj2).find(function (key) {
      return typeof obj1[key] === 'undefined';
    });
    if (diff) {
      return false;
    }
  }
  return obj1 === obj2;
}

function getValueByKey(resource, key) {
  var group = key.split('.');
  var currentObject = resource;
  var value = void 0;
  group.forEach(function (currentKey) {
    if (!currentObject[currentKey]) {
      return false;
    }
    if (isObject(currentObject[currentKey])) {
      currentObject = currentObject[currentKey];
    } else {
      value = currentObject[currentKey];
    }
    return true;
  });
  return value;
}

function setValueByKey(resource, key, value) {
  var group = key.split('.');
  var currentObject = resource;
  group.forEach(function (currentKey, index) {
    if (!currentObject[currentKey] && index < group.length - 1) {
      currentObject[currentKey] = {};
      currentObject = currentObject[currentKey];
    } else if (isObject(currentObject[currentKey])) {
      currentObject = currentObject[currentKey];
    } else if (group.length - 1 === index) {
      currentObject[currentKey] = value;
    }
  });
}

exports.default = { isObject: isObject, getValueByKey: getValueByKey, setValueByKey: setValueByKey };