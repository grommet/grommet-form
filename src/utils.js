export function isObject(item) {
  return (item && typeof item === 'object' && !Array.isArray(item));
}

export function deepMerge(target, ...sources) {
  if (!sources.length) {
    return target;
  }
  // making sure to not change target (immutable)
  const output = { ...target };
  const source = sources.shift();
  if (isObject(output) && isObject(source)) {
    Object.keys(source).forEach((key) => {
      if (isObject(source[key])) {
        if (!output[key]) {
          output[key] = { ...source[key] };
        } else {
          output[key] = deepMerge({}, output[key], source[key]);
        }
      } else {
        output[key] = source[key];
      }
    });
  }
  return deepMerge(output, ...sources);
}

export function getValueByKey(resource, key) {
  const group = key.split('.');
  let currentObject = resource;
  let value;
  group.forEach((currentKey) => {
    if (!currentObject[currentKey]) {
      return false;
    }
    if (isObject(currentObject[currentKey])) {
      currentObject = currentObject[currentKey];
    } else {
      value = currentObject[currentKey];
    }
    return true;
  });
  return value;
}

export function setValueByKey(resource, key, value) {
  const group = key.split('.');
  let currentObject = resource;
  group.forEach((currentKey, index) => {
    if (!currentObject[currentKey] && index < group.length - 1) {
      currentObject[currentKey] = {};
      currentObject = currentObject[currentKey];
    } else if (isObject(currentObject[currentKey])) {
      currentObject = currentObject[currentKey];
    } else if (group.length - 1 === index) {
      currentObject[currentKey] = value;
    }
  });
}

export default { isObject, getValueByKey, setValueByKey };
