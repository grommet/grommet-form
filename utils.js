'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.isObject = isObject;
exports.deepMerge = deepMerge;
exports.getValueByKey = getValueByKey;
exports.setValueByKey = setValueByKey;
function isObject(item) {
  return item && (typeof item === 'undefined' ? 'undefined' : _typeof(item)) === 'object' && !Array.isArray(item);
}

function deepMerge(target) {
  for (var _len = arguments.length, sources = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    sources[_key - 1] = arguments[_key];
  }

  if (!sources.length) {
    return target;
  }
  // making sure to not change target (immutable)
  var output = _extends({}, target);
  var source = sources.shift();
  if (isObject(output) && isObject(source)) {
    Object.keys(source).forEach(function (key) {
      if (isObject(source[key])) {
        if (!output[key]) {
          output[key] = _extends({}, source[key]);
        } else {
          output[key] = deepMerge({}, output[key], source[key]);
        }
      } else {
        output[key] = source[key];
      }
    });
  }
  return deepMerge.apply(undefined, [output].concat(sources));
}

function getValueByKey(resource, key) {
  var group = key.split('.');
  var currentObject = resource;
  var value = void 0;
  group.forEach(function (currentKey) {
    if (!currentObject[currentKey]) {
      return false;
    }
    if (isObject(currentObject[currentKey])) {
      currentObject = currentObject[currentKey];
    } else {
      value = currentObject[currentKey];
    }
    return true;
  });
  return value;
}

function setValueByKey(resource, key, value) {
  var group = key.split('.');
  var currentObject = resource;
  group.forEach(function (currentKey, index) {
    if (!currentObject[currentKey] && index < group.length - 1) {
      currentObject[currentKey] = {};
      currentObject = currentObject[currentKey];
    } else if (isObject(currentObject[currentKey])) {
      currentObject = currentObject[currentKey];
    } else if (group.length - 1 === index) {
      currentObject[currentKey] = value;
    }
  });
}

exports.default = { isObject: isObject, getValueByKey: getValueByKey, setValueByKey: setValueByKey };