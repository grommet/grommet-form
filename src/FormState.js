import { deepMerge, isObject, getValueByKey, setValueByKey } from './utils';

export default class FormState {
  constructor(rules, obj = {}, onChange) {
    if (!rules || (typeof rules !== 'function' && !Object.keys(rules).length)) {
      throw new Error('Rules is a required argument');
    }
    this.rules = rules;
    this.obj = obj;
    this.onChange = onChange;
  }
  set = (key, value) => {
    if (typeof key === 'string') {
      setValueByKey(this.obj, key, value);
    } else if (typeof key === 'object') {
      this.obj = deepMerge(this.obj, key);
    }

    if (this.onChange) {
      this.onChange();
    }
  };
  get = () => this.obj;
  getErrors = (rules, target) => {
    const errors = {};
    const targetRules = (
      rules || (typeof this.rules === 'function' ? this.rules(this.obj) : this.rules)
    );
    const targetObject = target || this.obj;
    Object.keys(targetRules).forEach((key) => {
      const rule = targetRules[key];
      if (typeof rule === 'function') {
        const value = getValueByKey(targetObject, key);
        if (Array.isArray(value)) {
          const errorsArray = [];
          value.forEach((v, index) => {
            const message = rule(value, index, targetObject);
            if (message) {
              errorsArray[index] = message;
            }
          });
          if (errorsArray.length) {
            setValueByKey(errors, key, errorsArray);
          }
        } else {
          const message = rule(value, undefined, targetObject);
          if (message) {
            errors[key] = message;
          }
        }
      } else if (isObject(rule)) {
        const ruleErrors = this.getErrors(rule, targetObject[key]);
        if (Object.keys(ruleErrors).length) {
          errors[key] = ruleErrors;
        }
      } else if (typeof rule === 'string' && !targetObject[key]) {
        setValueByKey(errors, key, rule);
      }
    });
    return errors;
  };
  isValid = () => !Object.keys(this.getErrors()).length
}
