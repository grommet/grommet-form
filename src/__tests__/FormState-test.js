import FormState from '../FormState';

import { deepMerge } from '../utils';

test('FormState fails to instantiate without rules', () => {
  expect(() => new FormState()).toThrowError('Rules is a required argument');
  expect(() => new FormState({})).toThrowError('Rules is a required argument');
});

test('FormState instantiate with valid basic rule', () => {
  const formState = new FormState({
    email: 'is required',
  });
  expect(formState).toMatchSnapshot();
  expect(formState.isValid()).toBeFalsy();
  expect(formState.getErrors()).toMatchSnapshot();
});

const emailExpression = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

test('FormState validates rules', () => {
  const formState = new FormState((user) => {
    const defaultValidation = {
      email: (value) => {
        if (!value || value === '') {
          return 'Email is required';
        } else if (!emailExpression.test(value)) {
          return 'Email is not valid';
        }
        return undefined;
      },
      name: 'Name is required',
      size: 'Size is required',
      confirm: 'Please confirm answers',
      kids: (value, index) => {
        if (index >= 0) {
          if (!value[index].name) {
            return { name: 'Name is required' };
          }
        } else if (!value || !value.length) {
          return 'You need to have at least one kid';
        }
        return undefined;
      },
      address: {
        home: {
          street: 'Street is required',
        },
      },
    };
    if (
      user.address &&
      user.address.home &&
      user.address.home.street &&
      user.address.home.street !== ''
    ) {
      return deepMerge(defaultValidation, {
        address: {
          city: 'City is required if you provided street',
        },
      });
    }
    return defaultValidation;
  });
  expect(formState).toMatchSnapshot();
  expect(formState.isValid()).toBeFalsy();
  expect(formState.getErrors()).toMatchSnapshot();

  formState.set('email', 'a');

  expect(formState.isValid()).toBeFalsy();
  expect(formState.getErrors()).toMatchSnapshot();

  formState.set('kids', [{ age: 10 }]);

  expect(formState.isValid()).toBeFalsy();
  expect(formState.getErrors()).toMatchSnapshot();

  formState.set('address.home.street', 'street');

  expect(formState.isValid()).toBeFalsy();
  expect(formState.getErrors()).toMatchSnapshot();

  formState.set('email', 'a@test.com');
  formState.set('name', 'name');
  formState.set('size', 'big');
  formState.set('confirm', true);
  formState.set('address.city', 'city');
  formState.set('kids', [{ name: 'paul', age: 10 }]);

  expect(formState.isValid()).toBeTruthy();
  expect(formState.getErrors()).toMatchSnapshot();
  expect(formState.get()).toMatchSnapshot();
});

test('FormState validates empty array', () => {
  const formState = new FormState({
    kids: 'Provide Kids',
  });

  expect(formState.isValid()).toBeFalsy();
  expect(formState.getErrors()).toMatchSnapshot();

  formState.set('kids', []);

  expect(formState.isValid()).toBeFalsy();
  expect(formState.getErrors()).toMatchSnapshot();

  formState.set('kids', ['one kid']);

  expect(formState.isValid()).toBeTruthy();
  expect(formState.getErrors()).toMatchSnapshot();
  expect(formState.get()).toMatchSnapshot();
});

test('FormState validates empty string', () => {
  const formState = new FormState({
    name: 'is required',
  });
  expect(formState.isValid()).toBeFalsy();
  formState.set('name', '   ');
  expect(formState.isValid()).toBeFalsy();
  formState.set('name', '  test ');
  expect(formState.isValid()).toBeTruthy();
  expect(formState.get()).toMatchSnapshot();
  expect(formState.getResource()).toMatchSnapshot();
});
