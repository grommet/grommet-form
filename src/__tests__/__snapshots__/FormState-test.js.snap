// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormState instantiate with valid basic rule 1`] = `
FormState {
  "get": [Function],
  "getErrors": [Function],
  "isValid": [Function],
  "obj": Object {},
  "onChange": undefined,
  "rules": Object {
    "email": "is required",
  },
  "set": [Function],
}
`;

exports[`FormState instantiate with valid basic rule 2`] = `
Object {
  "email": "is required",
}
`;

exports[`FormState validates empty array 1`] = `
Object {
  "kids": "Provide Kids",
}
`;

exports[`FormState validates empty array 2`] = `
Object {
  "kids": "Provide Kids",
}
`;

exports[`FormState validates empty array 3`] = `Object {}`;

exports[`FormState validates empty array 4`] = `
Object {
  "kids": Array [
    "one kid",
  ],
}
`;

exports[`FormState validates empty string 1`] = `
Object {
  "name": "test",
}
`;

exports[`FormState validates rules 1`] = `
FormState {
  "get": [Function],
  "getErrors": [Function],
  "isValid": [Function],
  "obj": Object {},
  "onChange": undefined,
  "rules": [Function],
  "set": [Function],
}
`;

exports[`FormState validates rules 2`] = `
Object {
  "address": Object {
    "home": Object {
      "street": "Street is required",
    },
  },
  "confirm": "Please confirm answers",
  "email": "Email is required",
  "kids": "You need to have at least one kid",
  "name": "Name is required",
  "size": "Size is required",
}
`;

exports[`FormState validates rules 3`] = `
Object {
  "address": Object {
    "home": Object {
      "street": "Street is required",
    },
  },
  "confirm": "Please confirm answers",
  "email": "Email is not valid",
  "kids": "You need to have at least one kid",
  "name": "Name is required",
  "size": "Size is required",
}
`;

exports[`FormState validates rules 4`] = `
Object {
  "address": Object {
    "home": Object {
      "street": "Street is required",
    },
  },
  "confirm": "Please confirm answers",
  "email": "Email is not valid",
  "kids": Array [
    Object {
      "name": "Name is required",
    },
  ],
  "name": "Name is required",
  "size": "Size is required",
}
`;

exports[`FormState validates rules 5`] = `
Object {
  "address": Object {
    "city": "City is required if you provided street",
  },
  "confirm": "Please confirm answers",
  "email": "Email is not valid",
  "kids": Array [
    Object {
      "name": "Name is required",
    },
  ],
  "name": "Name is required",
  "size": "Size is required",
}
`;

exports[`FormState validates rules 6`] = `Object {}`;

exports[`FormState validates rules 7`] = `
Object {
  "address": Object {
    "city": "city",
    "home": Object {
      "street": "street",
    },
  },
  "confirm": true,
  "email": "a@test.com",
  "kids": Array [
    Object {
      "age": 10,
      "name": "paul",
    },
  ],
  "name": "name",
  "size": "big",
}
`;
